File: tools/supabase_tool.py
"""Supabase Database Persistence.

Writes scored account research to the account_research table.
Requires SUPABASE_URL and SUPABASE_SERVICE_KEY environment variables.
"""

import os
from supabase import create_client, Client
from claude_agent_sdk import tool

# Initialize Supabase client (lazy — only connects when first used)
_client: Client | None = None


def _get_client() -> Client:
    global _client
    if _client is None:
        url = os.environ.get("SUPABASE_URL")
        key = os.environ.get("SUPABASE_SERVICE_KEY")
        if not url or not key:
            raise ValueError("SUPABASE_URL and SUPABASE_SERVICE_KEY must be set")
        _client = create_client(url, key)
    return _client


@tool
def persist_to_db(account_data: dict) -> dict:
    """Write research results to Supabase.

    Args:
        account_data: Complete account object with score, research, and recommendations.
                      Expected fields: id/domain, company, domain, score, confidence,
                      rationale, key_evidence, recommended_persona, message_angle,
                      raw_research.

    Returns:
        Confirmation with the database row ID.
    """
    try:
        client = _get_client()

        row = {
            "account_id": account_data.get("id", account_data.get("domain")),
            "company_name": account_data.get("company", account_data.get("company_name")),
            "company_domain": account_data.get("domain", account_data.get("company_domain")),
            "score": account_data.get("score"),
            "confidence": account_data.get("confidence"),
            "rationale": account_data.get("rationale"),
            "key_evidence": account_data.get("key_evidence"),
            "recommended_persona": account_data.get("recommended_persona"),
            "message_angle": account_data.get("message_angle"),
            "raw_research": account_data.get("raw_research"),
            "status": "completed",
        }

        # Upsert — insert or update if account_id already exists
        result = client.table("account_research").upsert(row).execute()

        if result.data:
            return {"status": "persisted", "id": result.data[0]["id"]}
        else:
            return {"status": "error", "message": "No data returned from upsert"}

    except Exception as e:
        return {"status": "error", "message": str(e)}

File: src/tools/supabaseTool.ts
import { createClient, type SupabaseClient } from "@supabase/supabase-js";

import { OpenAITool } from "./openaiTools.js";

let client: SupabaseClient | null = null;

function getClient(): SupabaseClient {
  if (!client) {
    const url = process.env.SUPABASE_URL;
    const key = process.env.SUPABASE_SERVICE_KEY;
    if (!url || !key) {
      throw new Error("SUPABASE_URL and SUPABASE_SERVICE_KEY must be set");
    }
    client = createClient(url, key);
  }
  return client;
}

export async function persistToDb(
  accountData: Record<string, unknown>
): Promise<Record<string, unknown>> {
  try {
    const db = getClient();
    const row = {
      account_id: accountData.id ?? accountData.domain,
      company_name: accountData.company ?? accountData.company_name,
      company_domain: accountData.domain ?? accountData.company_domain,
      score: accountData.score,
      confidence: accountData.confidence,
      scorecard: accountData.scorecard,
      rationale: accountData.rationale,
      top_signals: accountData.top_signals,
      cap_applied: accountData.cap_applied,
      recommended_persona: accountData.recommended_persona,
      message_angle: accountData.message_angle,
      raw_research: accountData.raw_research,
      status: "completed"
    };

    const { data, error } = await db.from("account_research").upsert(row).select("id");
    if (error) {
      return { status: "error", message: error.message };
    }
    if (data && data.length > 0) {
      return { status: "persisted", id: data[0].id };
    }
    return { status: "error", message: "No data returned from upsert" };
  } catch (error) {
    return {
      status: "error",
      message: error instanceof Error ? error.message : String(error)
    };
  }
}

export const persistToDbTool = new OpenAITool({
  name: "persist_to_db",
  description: "Persist a scored account research record to Supabase.",
  parameters: {
    type: "object",
    properties: {
      account_data: {
        type: "object",
        description: "The complete scored account research object."
      }
    },
    required: ["account_data"],
    additionalProperties: false
  },
  handler: (args) => persistToDb(args.account_data as Record<string, unknown>)
});

export interface StakeholderRow {
  account_id: string;
  account_uuid?: string | null;
  company_name?: string | null;
  company_domain?: string | null;
  amplemarket_id: string;
  full_name?: string | null;
  title?: string | null;
  linkedin_url?: string | null;
  persona: string;
  sub_persona?: string | null;
  department: string;
  rank: string;
  decision_role?: string | null;
  confidence?: number | null;
  rationale?: string | null;
  evidence?: unknown;
  raw_input?: unknown;
  model?: string | null;
}

export async function persistStakeholders(
  rows: StakeholderRow[]
): Promise<{ status: "persisted" | "error"; inserted?: number; message?: string }> {
  if (rows.length === 0) return { status: "persisted", inserted: 0 };
  try {
    const db = getClient();
    const { data, error } = await db
      .from("account_stakeholders")
      .upsert(rows, { onConflict: "account_id,amplemarket_id,persona" })
      .select("id");
    if (error) return { status: "error", message: error.message };
    return { status: "persisted", inserted: data?.length ?? 0 };
  } catch (error) {
    return {
      status: "error",
      message: error instanceof Error ? error.message : String(error)
    };
  }
}
